src/Npgsql/Internal/TypeHandlers/ArrayHandler.cs

```diff
@@ -201,72 +196,91 @@ protected async ValueTask<List<TRequestedElement>> ReadList<TRequestedElement>(N
         internal static class ElementTypeInfo<TElement>
         {
             public static readonly bool IsNonNullable =
-                typeof(TElement).IsValueType && Nullable.GetUnderlyingType(typeof(TElement)) is null;
+                typeof(TElement).IsValueType && default(TElement) != null;
 
             public static readonly Type NullableElementType = IsNonNullable
                 ? typeof(Nullable<>).MakeGenericType(typeof(TElement))
                 : typeof(TElement);
         }
 
-        internal static class ArrayTypeInfo<TArrayOrList>
+        internal abstract class ArrayTypeInfo<TArray>
         {
             // ReSharper disable StaticMemberInGenericType
-            public static readonly bool IsArray;
-            public static readonly bool IsList;
-            public static readonly Type? ElementType;
-
-            public static readonly Func<ArrayHandler, NpgsqlReadBuffer, bool, ValueTask<Array>> ReadArrayFunc = default!;
-            public static readonly Func<ArrayHandler, NpgsqlReadBuffer, bool, ValueTask<TArrayOrList>> ReadListFunc = default!;
+            public static readonly Type? ElementType = typeof(TArray).IsArray ? typeof(TArray).GetElementType() : null;
             // ReSharper restore StaticMemberInGenericType
 
-            public static bool IsArrayOrList => IsArray || IsList;
+            public static bool IsArray => ElementType is not null;
 
-            static ArrayTypeInfo()
+            static ArrayTypeInfo<TArray>? _derivedInstance;
+            static ArrayTypeInfo<TArray> DerivedInstance
             {
-                var type = typeof(TArrayOrList);
-                IsArray = type.IsArray;
-                IsList = type.IsGenericType && type.GetGenericTypeDefinition() == typeof(List<>);
+                get
+                {
+                    return _derivedInstance ?? CreateInstance();
+                    static ArrayTypeInfo<TArray> CreateInstance()
+                    {
+                        if (ElementType is null)
+                            return null!;
+
+                        _derivedInstance = (ArrayTypeInfo<TArray>?)Activator.CreateInstance(typeof(ElementInfo<>).MakeGenericType(typeof(TArray), ElementType), typeof(TArray).GetArrayRank());
```

Comment:
Uh, isn't this exactly the runtime code generation we're trying to avoid?

Reply:
As discussed offline, there probably isn't a way to completely get rid of MakeGenericType/MakeGenericMethod entirely - at some point a `GetFieldValue<TArray>` needs to make the "jump" into an internal `GetValue<TElement>` (on the element's type handler), and that can't be done generically without some form of code generation. But you seem to show below that at least for the set of known types, that can be worked around with rd.xml.

Another option here would be to add a `GetArray<TElement>` or `GetList<TElement>` on NpgsqlDataReader, which would always work. That could be the answer if anyone needs arrays of "non-standard" types which aren't listed in rd.xml. It wouldn't be ADO.NET-standard, but it seems very acceptable at this point.

@NinoFloris please feel free to correct (or confirm!) any of the above, since I'm a bit of a newcomer to AOT/trimming and you obviously know what you're doing more than I.